# Weather Table Audit

## Objective

Perform a data quality assessment of the Weather dataset before loading it into the Silver Layer.

The audit focuses on:

- Missing values
- Duplicate records
- Primary key validation
- Data type validation
- Category standardization
- Business rule validation

In [1]:
import pandas as pd

## 1. Load Dataset

Load the Weather dataset into a Pandas DataFrame and perform an initial inspection.

In [2]:
weather = pd.read_csv(r"C:\Dataset\weather.csv")

## 2. Dataset Structure Review

Review:

- Number of records
- Number of columns
- Data types
- Overall dataset structure

In [3]:
weather.info

<bound method DataFrame.info of        weather_id            timestamp weather_type  temperature_c  \
0         W059771  2023-05-26 12:50:56         rain           20.6   
1         W021363     02-21-2023 12:49        clear           16.9   
2         W127325           2023/11/06         rain           38.9   
3         W140510    09-Dec-2023 01:02          clr           13.5   
4         W144298  2023-12-18 04:39:18         rain           26.7   
...           ...                  ...          ...            ...   
149995    W119880     10-19-2023 16:21        clear           21.6   
149996    W103695  2023-09-10 08:43:13        FOGGY           13.6   
149997    W131933  2023-11-17 20:38:32        Rainy           40.3   
149998    W146868    24-Dec-2023 08:05        clear           27.1   
149999    W121959       24/10/23 11:36         rain           12.3   

        visibility_km  
0            6.384281  
1                 NaN  
2            6.294870  
3            4.400000  
4      

In [4]:
weather.dtypes

weather_id        object
timestamp         object
weather_type      object
temperature_c    float64
visibility_km    float64
dtype: object

## 3. Missing Value Analysis

Identify missing values across all columns and evaluate their impact on data completeness.

In [5]:
weather.isnull().sum()

weather_id           0
timestamp            0
weather_type         0
temperature_c    10500
visibility_km    15000
dtype: int64

## 4. Duplicate Record Analysis

Check for fully duplicated records that may have resulted from ingestion or processing issues.

In [6]:
weather.duplicated().sum()

np.int64(0)

## 5. Primary Key Validation

Evaluate whether `weather_id` can serve as a reliable primary key by validating:

- NULL values
- Uniqueness
- Duplicate occurrences

In [7]:
weather['weather_id'].nunique()

145555

In [8]:
weather['weather_id'].duplicated().sum()

np.int64(4445)

## 6. Timestamp Validation

Review timestamp values for:

- Data type correctness
- Format consistency
- Potential conversion issues

The objective is to prepare the column for temporal validation and analysis.

In [9]:
weather['timestamp']

0         2023-05-26 12:50:56
1            02-21-2023 12:49
2                  2023/11/06
3           09-Dec-2023 01:02
4         2023-12-18 04:39:18
                 ...         
149995       10-19-2023 16:21
149996    2023-09-10 08:43:13
149997    2023-11-17 20:38:32
149998      24-Dec-2023 08:05
149999         24/10/23 11:36
Name: timestamp, Length: 150000, dtype: object

## 7. Weather Type Standardization Review

Inspect weather categories for:

- Inconsistent casing
- Spelling variations
- Invalid category values

In [10]:
weather['weather_type'].unique()

array(['rain', 'clear', 'clr', 'RAIN', 'Clear', 'Rainy', 'Rain', 'rainy',
       'rain ', 'raining', 'RAINY', 'Clr', 'fog', 'clear ', 'CLEAR',
       'CLR', 'FOG', 'Fog', 'FOGGY', 'Fog ', 'foggy', 'Foggy'],
      dtype=object)

## 8. Temperature Validation

Validate temperature measurements against expected business rules.

The objective is to identify:

- Missing values
- Extreme values
- Potential outliers
- Data quality issues

In [11]:
weather[weather['temperature_c'] > 50].count()

weather_id       1651
timestamp        1651
weather_type     1651
temperature_c    1651
visibility_km    1482
dtype: int64

In [12]:
weather[weather['temperature_c'] < 0].count()

weather_id       1140
timestamp        1140
weather_type     1140
temperature_c    1140
visibility_km    1026
dtype: int64

## 9. Visibility Validation

Validate visibility measurements and identify values that may violate business rules.

The objective is to identify:

- Missing values
- Negative values
- Extreme values
- Potential data quality issues

In [13]:
weather[weather['visibility_km'] == 0].count()

weather_id       366
timestamp        366
weather_type     366
temperature_c    342
visibility_km    366
dtype: int64

In [14]:
weather[weather['visibility_km'] < 0].count()

weather_id       414
timestamp        414
weather_type     414
temperature_c    380
visibility_km    414
dtype: int64

## 10. Audit Findings Summary

### Key Findings

- 10,500 missing temperature values
- 15,000 missing visibility values
- 4,445 duplicate weather_id values
- Timestamp stored as object datatype
- Inconsistent weather_type categories
- 1,651 temperatures above 50°C
- 704 negative visibility values

### Candidate Primary Key Status

❌ weather_id cannot currently be used as a primary key.

### Recommended Silver Layer Actions

- Resolve duplicate weather_id values
- Handle missing values
- Convert timestamp to datetime format
- Standardize weather_type values
- Validate temperature ranges
- Validate visibility measurements